### PR TITLE
kmsv2: set authority to localhost

### LINKED ISSUE #126929: Encryption at rest KMS plugin receives invalid authority headers from k8s grpc client
### What happened?

KMS Plugins require communication via grpc UDS. Currently, the k8s grpc client sets the authority header to the socket path which is marked as invalid preventing successful socket communication. 

### What did you expect to happen?

I expect a custom KMS provider plugin to be sent valid authority headers so that successful communication can be established between the k8s and grpc client/server.

### How can we reproduce it (as minimally and precisely as possible)?

Create a grpc server according to the [k8s documentation on kms v2](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/#configuring-the-kms-provider-kms-v2) and attempt to establish socket communication.

### Anything else we need to know?

This issue has been resolved in [this PR](https://github.com/kubernetes/kubernetes/pull/112597) for device plugins and CSI. The same solution will likely fix this issue as well. I will create a PR for this and link it here after.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.30.4
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.29.4+k3s1
```

</details>

### Cloud provider

<details>
None
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy

$ uname -a
Linux computer-XPS-15-9530 6.5.0-44-generic #44~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Jun 18 14:36:16 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by Ruddickmg**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
KMS Plugins require communication via grpc UDS. Currently, the k8s grpc client sets the authority header to the socket path which is marked as invalid preventing successful socket communication.

This issue was addressed in other parts of the code base but not here. Since the only way to communicate with KMSv2 according to the [documentation](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/#configuring-the-kms-provider-kms-v2) is via UDS and grpc, it seems essential that this fix should be applied here as well.

#### Which issue(s) this PR fixes:
Fixes [#126929](https://github.com/kubernetes/kubernetes/issues/126929)

#### Special notes for your reviewer:
This fix is in line with previous solutions for an identical problem. [The previous solution can be found here](https://github.com/kubernetes/kubernetes/pull/112597)

#### Does this PR introduce a user-facing change?

```release-note
Fixed an issue where requests sent by the KMSv2 service would be rejected due to having an invalid authority header.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
```

### CODE REVIEW

**Comment by sftim** [comment]
Does this need a release note?

**Comment by dims** [comment]
/ok-to-test

/lgtm

here are other instances: https://cs.k8s.io/?q=grpc.WithAuthority%5C(&i=nope&files=&excludeFiles=vendor&repos=

/assign @liggitt @enj

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go:1]: Please revert this change.

KMS v1 is deprecated and off by default, and no one has ever complained about this issue in the many years that this problem has existed.  We should not encourage any further integration with it by fixing issues, especially since the long term plan is to delete it altogether (timeline TBD).

Outside of critical security fixes, no changes will be made to KMS v1.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ func NewGRPCService(ctx context.Context, endpoint, providerName string, callTime
 	s := &gRPCService{callTimeout: callTimeout}
 	s.connection, err = grpc.Dial(
 		addr,
+		grpc.WithAuthority("localhost"),
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go:61]: Can we check for this in the KMS v2 integration tests?

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go
Comment (Ruddickmg) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go:1]: I was not using KMS v1, just thought it might make sense to make the change in both places, will revert.

Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service.go:1]: My understanding from the linked issue is that this has always been broken when using a gRPC server implementation not written in Go (because they try to parse the authority value and fail).

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go (hunk 1 above)
Comment (Ruddickmg) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go:61]: I'm not too familiar with golang so doing my best. It looks like all the grpc code is imported from [google.golang.org/grpc](https://pkg.go.dev/google.golang.org/grpc) which is an external project so I'm not sure how to check the header of the grpc request in a go test, or how to simulate sending a request to a grpc service not written in go from a go test 😅. I could possibly try to intercept the request from the socket and see if I could check the header there?

FWIW I noticed that in the [PR for fixing the other areas of code with this problem](https://github.com/kubernetes/kubernetes/pull/112597/files) there were not tests modified or created.

Is this a blocker?

Comment (enj) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go:61]:
```patch
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go b/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go
index 2babbbe3c7d..ee6b9ef1cca 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go
@@ -31,6 +31,7 @@ import (
 
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
 	"google.golang.org/grpc/status"
 
 	"k8s.io/apimachinery/pkg/util/wait"
@@ -61,7 +62,16 @@ type Base64Plugin struct {
 
 // NewBase64Plugin is a constructor for Base64Plugin.
 func NewBase64Plugin(t testing.TB, socketPath string) *Base64Plugin {
-	server := grpc.NewServer()
+	server := grpc.NewServer(
+		grpc.UnaryInterceptor(
+			func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+				if val := metadata.ValueFromIncomingContext(ctx, ":authority"); len(val) != 1 || val[0] != "localhost" {
+					t.Errorf("wanted localhost authority, got: %v", val)
+				}
+				return handler(ctx, req)
+			},
+		),
+	)
 	result := &Base64Plugin{
 		grpcServer: server,
 		mu:         &sync.Mutex{},
```

Comment (Ruddickmg) [reply staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go:61]: @enj Thank you! Added

**Comment by enj** [comment]
@Ruddickmg please squash to one commit and add a release note to describe the bug fix.

**Comment by Ruddickmg** [comment]
> @Ruddickmg please squash to one commit and add a release note to describe the bug fix.

I attempted to squash, then it wouldn't let me push the squashed commit, my IDE suggested merging, and now I think I've messed up the git history in a way that I'm not sure how to fix. Should I just create a new PR?
